File: circularqueue.c
#include <stdio.h>
#include <stdlib.h>
#define SIZE 5
typedef struct {
    int items[SIZE];
    int front;
    int rear;
} CircularQueue;
void initialize(CircularQueue *q);
int isFull(CircularQueue *q);
int isEmpty(CircularQueue *q);
void enqueue(CircularQueue *q, int value);
int dequeue(CircularQueue *q);
int peek(CircularQueue *q);
void display(CircularQueue *q);

int main() {
    CircularQueue q;
    initialize(&q);

    int choice, value;

    while (1) {
        printf("\nCircular Queue Menu:\n");
        printf("1. Enqueue\n");
        printf("2. Dequeue\n");
        printf("3. Peek\n");
        printf("4. Display\n");
        printf("5. Exit\n");
        printf("Enter your choice: ");
        scanf("%d", &choice);

        switch (choice) {
            case 1:
                printf("Enter value to enqueue: ");
                scanf("%d", &value);
                enqueue(&q, value);
                break;
            case 2:
                dequeue(&q);
                break;
            case 3:
                value = peek(&q);
                if (value != -1)
                    printf("Front element: %d\n", value);
                break;
            case 4:
                display(&q);
                break;
            case 5:
                printf("Exiting program.\n");
                exit(0);
            default:
                printf("Invalid choice. Try again.\n");
        }
    }

    return 0;
}
void initialize(CircularQueue *q) {
    q->front = -1;
    q->rear = -1;
}
int isFull(CircularQueue *q) {
    return ((q->rear + 1) % SIZE == q->front);
}
int isEmpty(CircularQueue *q) {
    return (q->front == -1);
}
void enqueue(CircularQueue *q, int value) {
    if (isFull(q)) {
        printf("Queue is full! Cannot enqueue %d\n", value);
        return;
    }

    if (isEmpty(q)) {
        q->front = 0;
    }

    q->rear = (q->rear + 1) % SIZE;
    q->items[q->rear] = value;
    printf("Enqueued %d\n", value);
}
int dequeue(CircularQueue *q) {
    if (isEmpty(q)) {
        printf("Queue is empty! Cannot dequeue\n");
        return -1;
    }

    int value = q->items[q->front];

    if (q->front == q->rear) {
        q->front = -1;
        q->rear = -1;
    } else {
        q->front = (q->front + 1) % SIZE;
    }

    printf("Dequeued %d\n", value);
    return value;
}
int peek(CircularQueue *q) {
    if (isEmpty(q)) {
        printf("Queue is empty! Nothing to peek\n");
        return -1;
    }

    return q->items[q->front];
}
void display(CircularQueue *q) {
    if (isEmpty(q)) {
        printf("Queue is empty!\n");
        return;
    }

    printf("Queue contents: ");
    int i = q->front;
    while (1) {
        printf("%d ", q->items[i]);
        if (i == q->rear)
            break;
        i = (i + 1) % SIZE;
    }
    printf("\n");
}

File: Circularqueue.java
/*
 * Click nbfs://nbhost/SystemFileSystem/Templates/Licenses/license-default.txt to change this license
 */

package winnie.circularqueque;

/**
 *
 * @author Winnie
 */
public class Circularqueue{
private int[]queue;
private int front,rear,size,capacity;
public circularqueue(int capacity){
    this.capacity=capacity;
    this.queue=new int[capacity];
    this.front=0;
    this rear=-1;
    this size=0;
}
public boolean enqueue(int value){
    if(isFull()){
    return false;//queque full
}
    rear=(rear+1)%capacity;
    queue[rear]=value;
    size--;
    return true;
}
public boolean dequeue(){
    if(isEmpty()){
        return false;//queue empty
    }
    front=(front+1)%capacity;
    size--;
    return true;
}
public boolean isFull(){
    return size==capacity;
}
public boolean isEmpty(){
    return size==0;
}
public void display(){
    if(isEmpty()){
        system.out.println("queue is empty");
    }
    system.out.print.ln("queue");
    for(int i=0;i=size,i++){
        int index=(front+i)%capacity;
        system.out.ln(queue[index]+" ");
    }
    system.out.println();
}
public static void main(string[]args){
    circular queue cq=new circularqueue(3);
    system.out.println(cq.enqueue(10));
    system.out.println(cq.enqueue(20));
    system.out.println(cq.enqueue(30));
    system.out.println(cq.enqueue(40));
    cq.display();
    system.out.println(cq.dequeue());
    cq.display();
    system.out.println(cq.enqueue(40));
    cq.display();
    system.out.println("is full?"+cq.isFull());
}
}
